Linksys

mellowCS opened the pull request.

mellowCS committed:
added download functionality for linksys firmware

mellowCS committed:
added functionality to linksys.py and added fields to FirmwareItem in items.py

mellowCS committed:
added test and mock file
removed check method for firmware urls and integrated check into xpath

mellowCS committed:
added a custom files pipeline for the linksys crawler
and changed the correspondent reference in the settings

mellowCS committed:
added better version and date extraction to the crawler

mellowCS committed:
- Finished test_linksys.py
- Generator return type hints added to the corresponding methods

- Bug Fixes:
	- when regex does not result in a match in the prepare_meta_data method, an error occurs (fix: N/A ist applied to the item key)
	- One missing entry in the device class dictionary led to an UnknownDeviceClassException (fix: Entry has been added to the dictionary)
	- UnknownDeviceClassException was not properly handled (fixed)

mellowCS committed:
refactored parse firmware into two methods for less nesting
added corresponding tests

dorpvom commented:
Please always remove .pyc files from PRs.
Best way is to add a .gitignore file. This will keep you from adding specific (kinds) of files from to commits in the first place.

dorpvom commented on firmware/pipelines.py:
```diff
@@ -5,7 +5,14 @@
+from scrapy.pipelines.files import FilesPipeline
+
 
 class FirmwarePipeline(object):
```
Is this still needed?

mellowCS replied:
All spiders have to share the items.py, settings.py, pipeline.py and middlewares.py in the final project. I don't know if any other spider will use the FIrmwarePipeline class in the future but mine won't need it as it uses the LinksysPipeline.

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+        version = match.group(1).strip(' ') if match else 'N/A'
+
+        match = re.search(r'([^\d]+)(\d+\/\d+\/\d{4}).*', firmware)
+        date = self.convert_date(match.group(2)) if match else 'N/A'
```
This is still in german **DD-MM-YYYY** notation. Should be **YYYY-MM-DD**

Consider using some function from `datetime.datetime` or `time` packages

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+    device_classes = {ClassIdentifier(['MR', 'EA', 'WRT', 'E', 'BEF', 'WKU', 'WRK']): 'Router',
+                      ClassIdentifier(['X', 'AG', 'WAG']): 'Modem Router',
+                      ClassIdentifier(['LRT']): 'VPN Router', ClassIdentifier(['WHW', 'VLP']): 'Wifi Mesh System',
+                      ClassIdentifier(['LAPN', 'LAPAC']): 'Business Access Point', ClassIdentifier(['WAP']): 'Home Access Point',
```
Reformat to one class per line

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+
+
+class LinksysSpider(Spider):
+    handle_httpstatus_list = [404]
```
class globals that are defined outside of __init__ should be all uppercase by pep convention

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+from scrapy import Request, Spider
```
Refactor import order.
PEP style is: built ins first (i.e. re, typing), third party second (i.e. scrapy), first party third (i.e. firmware.items).
Then alphabetical in each block plus a blank line dividing the blocks

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+        for version in response.xpath(self.x_path['firmware']).extract():
+            yield from self.parse_firmware(device_name=device_name, version=version)
+
+    def parse_firmware(self, device_name: str, version: str):
```
Consider refactoring into (at least) two functions

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+
+
+class UnknownDeviceClassException(Exception):
+    def __init__(self, product):
```
I'd think you can refactor this to 
```python
class UnknownDeviceClassException(Exception):
    pass
```
And then pass the complete error message below instead of only passing the product

dorpvom commented on firmware/spiders/linksys.py:
```diff
@@ -0,0 +1,111 @@
+        raise UnknownDeviceClassException(product=product)
+
+    @staticmethod
+    def convert_date(date: str) -> str:
```
see above

dorpvom commented on firmware/tests/test_linksys.py:
```diff
@@ -0,0 +1,178 @@
+        assert list(spider_instance.parse_versions(response=response, device_name=device_name))[0] == expected
+
+
+@pytest.mark.parametrize('device_name, version, expected', [('Broadband Router', FIRMWARE_PAGE,
```
Formatting looks off. Just linebreak after `[` and continue with 4 spaces offset

dorpvom commented on firmware/tests/test_linksys.py:
```diff
@@ -0,0 +1,178 @@
+
+
+@pytest.mark.parametrize('meta_data, expected', [(dict(
+                                                       file_urls='http://downloads.linksys.com/downloads/firmware/FW_EA6300_1.203.23.20394_prod.gpg.img',
```
see above (formatting)

dorpvom reviewed (changes requested):
Mostly fine. Just some minor changes necessary.

dorpvom committed:
Merge branch 'precommit' of https://github.com/mellowCS/FirmwareScraper into linksys

mellowCS committed:
removed convert_date method and replaced it with datetime functionality
made some minor cosmetic changes to the code and fixed a bug where
the regex for firmware urls would also extract additional tags

mellowCS committed:
Merge branch 'master' of https://github.com/mellowCS/FirmwareScraper into linksys

mellowCS committed:
Delete __init__.cpython-36.pyc

mellowCS committed:
Delete linksys.cpython-36.pyc

mellowCS committed:
Delete __init__.cpython-36.pyc

mellowCS committed:
Delete items.cpython-36.pyc

mellowCS committed:
Delete settings.cpython-36.pyc

mellowCS committed:
fixed import order in linksys.py

dorpvom committed:
Merge branch 'linksys' of https://github.com/mellowCS/FirmwareScraper into linksys

dorpvom committed:
test imports and cosmetic
